### DISCUSSION TITLE
Sparkfun Micromod STM32 - Add additional UART

### DISCUSSION THREAD

**Discussion by SYNiMAK** [micropython/micropython#14065] [STM32 / Pyboard] [answered]
I am looking to modify the SPARKFUN_MICROMOD_STM32 port of micropython to add a second uart on pins PB6 and PB7.

I have modified mpconfigboard.h and pins.csv but I am obviously missing something because when I compile and load the firmware I get a value error for "UART(2)"

This is what I added to mpconfigboard.h...

```
// UART2 config (MicroMod UART2)
#define MICROPY_HW_UART2_NAME "UART2"
#define MICROPY_HW_UART2_TX (pin_B6)
#define MICROPY_HW_UART2_RX (pin_B7)
```

and this to pins.csv...

```
UART_TX2, PB6
UART_RX2, PB7
```

I have compiled with these commands...
```
cd micropython/ports/stm32
make BOARD=SPARKFUN_MICROMOD_STM32 submodules
make BOARD=SPARKFUN_MICROMOD_STM32
```

What else do I have to change to get this to work?

Any help greatly appreciated!

**Comment by davefes** [comment]
Maybe start with:
`make BOARD=SPARKFUN_MICROMOD_STM32 clean`

**Reply by SYNiMAK** [reply]
Yeah tried a clean. Didn't make any difference. 

But I did find  `pins_SPARKFUN_MICROMOD_STM32.c` which is generated by `make-pins.py`. That seems to have the answers I am looking for. Giving it a go today.

**Reply by davefes** [reply]
Ok, when I enabled a 2nd UART on a STM32F411 I did not change any pins.csv.

**Reply by SYNiMAK** [reply, marked as answer]
I have UART1, UART3 and UART6 working. 

Looking through 'pins_SPARKFUN_MICROMOD_STM32.c' gave me the pins that the different uart ports work on. 

Modified 'pins_SPARKFUN_MICROMOD_STM32.c' and 'pins.csv' and it works.

**Reply by davefes** [reply]
Could you point to a guide that describes how to make these changes properly?

**Reply by SYNiMAK** [reply]
@davefes I didn't really follow any guide. There is probable a much better way to do it from my hack it together way but the general steps were...

View the pins_SPARKFUN_MICROMOD_STM32.c file and work out which pins can be used with which UART ports.
This may require a compile first to generate the file.

Modify pins.csv and mpconfigboard.h to reflect the changes in pins that you want to make

Compile and test

**Comment by davefes** [comment]
In micropython/ports.stm32/SPARKFUN_MICROMOD_STM32 I see a pins.csv, which seems more descriptive than the pins.csv I have seen so far ... which isn't many!

Where is the file `pins_SPARKFUN_MICROMOD_STM32.c` ??  

I have been trying to put Micropython on a E77-400M22S.  I think I managed to get a 2nd UART working, but can't sure.

Thanks,
Dave
